src/Configuration.php

```diff
@@ -1823,4 +1823,34 @@ private static function looksLikeAPipe($stream): bool
 
         return $mode === 0010000 || $mode === 0040000 || $mode === 0100000 || $mode === 0120000;
     }
+
+    /**
+     * Locate a command (an executable) in $PATH.
+     *
+     * Behaves like 'command -v COMMAND' or 'which COMMAND'.
+     * If $PATH is unset/empty it defaults to '/usr/sbin:/usr/bin:/sbin:/bin'.
+     *
+     * @param string $command the executable to locate
+     *
+     * @return string
+     */
+    private static function which($command)
+    {
+        $path = \getenv('PATH');
+        if (empty($path)) {
+            $path = '/usr/sbin:/usr/bin:/sbin:/bin';
+        }
+        $paths = \explode(':', $path);
+        foreach ($paths as $path) {
+            $fullpath = "$path/$command";
+            $st = @\stat($fullpath);
+            if ($st !== false &&
+                    ($st['mode'] & 0170000) === 0100000 &&   // S_IFREG
+                    ($st['mode'] & 0555) !== 0) {            // any executable bit
```

Comment:
i threw together a simple test fixture and it `$st['mode'] & 0555` seems to be true for regular files even with no executable bit set. i do get the expected result for `is_executable`.

it looks like `is_file`, `is_link` and `is_executable` uses `stat()` under the hood (and cache the result) so should we use those directly instead? 

```
if ((\@is_file($fullpath) || \@is_link($fullpath)) && \@is_executable($fullpath)) {
    return $fullpath;
}
```

Reply:
Ha. I threw myself off there.
`($st['mode'] & 0111) !== 0)` was the original idea, but I figured we need to be able to read the executables too. (And then I forgot that we couldn't do a simple boolean compare anymore.)

But I guess we don't need `04`:
```
$ ls -l /usr/bin/zip
---x--x--x 1 root root 203768 mrt 25 10:55 /usr/bin/zip
```
```
$ which zip
/usr/bin/zip
```
```
$ zip
Copyright (c) 1990-2008 Info-ZIP - Type 'zip "-L"' for software license.
...
```
So that should just be `0111` then. 

But for anything other than `0001` the user/group needs to be checked too: so you'd need an `access(..., X_OK)` call as well.

`which` uses `newfstatat` + `faccessat2(..., X_OK, ...)`.
`is_file` uses `newfstatat` and `is_executable` uses `access(..., X_OK)`, so that is a perfect match.

Fixing...

Reply:
> But I guess we don't need 04:

Okay, we would, if it is a script. But we don't need to be smarter than `which` here.
```
$ which less
/home/walter/bin/less
```
```
$ less
/bin/sh: 0: cannot open /home/walter/bin/less: Permission denied
```
Ignoring that...